src/groups.py

```diff
@@ -0,0 +1,380 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper class used to manage interactions with the Ranger API for user/group operations."""
+
+import copy
+import json
+import logging
+
+import requests
+import yaml
+from apache_ranger.exceptions import RangerServiceException
+from ops.charm import CharmBase
+from ops.model import ActiveStatus
+
+from literals import (
+    ADMIN_USER,
+    ENDPOINT_MAPPING,
+    EXPECTED_KEYS,
+    HEADERS,
+    MEMBER_TYPE_MAPPING,
+    RANGER_URL,
+    SYSTEM_GROUPS,
+)
+from utils import (
+    generate_random_string,
+    handle_service_error,
+    log_event_handler,
+    retry,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RangerGroupManager:
+    """This class handles the communication with the Ranger API."""
+
+    def __init__(self, charm: CharmBase):
+        """Construct RangerGroupManager object.
+
+        Args:
+            charm: the charm for which this relation is provided
+        """
+        self.charm = charm
+
+    @property
+    def _auth(self):
+        """Return the authentication credentials."""
+        admin_password = self.charm.config["ranger-admin-password"]
+        auth = (ADMIN_USER, admin_password)
+        return auth
+
+    @log_event_handler(logger)
+    def _handle_synchronize_file(self, event):
+        """Synchronize the data from the config file with the Ranger API.
+
+        Args:
+            event: The config changed event that triggered the synchronization.
+        """
+        if not self.charm.unit.status == ActiveStatus("Status check: UP"):
+            logger.debug("Status not active, event deferred.")
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        config_data = self._read_file()
+        for key in config_data:
+            data = next(iter(config_data.values()))
+            relation_data = copy.deepcopy(data)
+            self._synchronize(key, data, event)
+            self._add_to_relations(key, relation_data)
+
+    def _read_file(self):
+        """Read the user-group-configuration from the charm config.
+
+        Returns:
+            Data read from the configuration file.
+        """
+        config = self.charm.config["user-group-configuration"]
+        data = yaml.safe_load(config)
+        return data
+
+    def _synchronize(self, key, data, event):
+        """Synchronize data with the Ranger API.
+
+        Args:
+            key: relation_id from configuration file.
+            data: Data to synchronize with the Ranger API.
+            event: configuration changed event.
+        """
+        try:
+            self._sync(data["groups"], "group")
```

Comment:
maybe we could use the same strings here (groups instead of group and so on), and maybe pass (data, member_type) and then do config = data[member_type] ... just a suggestion to reduce stuttering.. not really important

Reply:
I am planning to turn this into a library and combine some methods from the trino and ranger charm. So as it does not change the functionality, I will implement this then if that is ok.